File: swift/Sources/OpenID4VCI/Crypto.swift
import CborCose
import Crypto
import Foundation
import SdJwt
import WalletAPI

func sha256(_ bytes: [UInt8]) -> [UInt8] {
    [UInt8](SHA256.hash(data: Data(bytes)))
}

func jwsAlgName(_ alg: SigningAlgorithm) -> String {
    switch alg {
    case .es256: return "ES256"
    case .es384: return "ES384"
    case .es512: return "ES512"
    }
}

/// PKCE (RFC 7636) with the S256 method — the only method HAIP permits.
public struct Pkce {
    public let codeVerifier: String
    public let codeChallenge: String
    public var method: String { "S256" }

    public static func create(rng: any Rng) -> Pkce {
        let verifier = Base64Url.encode(rng.nextBytes(32))
        let challenge = Base64Url.encode(sha256([UInt8](verifier.utf8)))
        return Pkce(codeVerifier: verifier, codeChallenge: challenge)
    }
}

/// DPoP proof builder (RFC 9449). Produces a fresh proof per request (unique jti, correct
/// htm/htu, optional ath binding to the access token, and a server nonce on retry).
public struct DpopProver {
    private let signer: any JwsSigner
    private let jwk: JsonValue
    private let rng: any Rng
    private let now: () -> Int64

    public init(signer: any JwsSigner, publicKey: EcPublicKey, rng: any Rng, now: @escaping () -> Int64) {
        self.signer = signer
        self.jwk = JwkEc.toJson(publicKey)
        self.rng = rng
        self.now = now
    }

    public func proof(method: String, url: String, accessToken: String? = nil, nonce: String? = nil) async throws -> String {
        let header = JsonValue.obj([
            ("typ", .str("dpop+jwt")),
            ("alg", .str(jwsAlgName(signer.algorithm))),
            ("jwk", jwk),
        ])
        var claims: [(String, JsonValue)] = [
            ("jti", .str(Base64Url.encode(rng.nextBytes(16)))),
            ("htm", .str(method)),
            ("htu", .str(htu(url))),
            ("iat", .numInt(now())),
        ]
        if let accessToken {
            claims.append(("ath", .str(Base64Url.encode(sha256([UInt8](accessToken.utf8))))))
        }
        if let nonce {
            claims.append(("nonce", .str(nonce)))
        }
        let jws = try await Jws.sign(header: header, payload: [UInt8](JsonValue.obj(claims).serialize().utf8), signer: signer)
        return jws.compact()
    }

    /// htu is the request URI without query and fragment (RFC 9449 §4.2).
    private func htu(_ url: String) -> String {
        var s = url
        if let q = s.firstIndex(of: "?") { s = String(s[s.startIndex..<q]) }
        if let f = s.firstIndex(of: "#") { s = String(s[s.startIndex..<f]) }
        return s
    }
}

/// OpenID4VCI key-proof-of-possession builder (§8.2.1): typ `openid4vci-proof+jwt`, the
/// holder's public JWK in the header, `aud`=issuer, and the issuer `c_nonce`.
public struct KeyProofSigner {
    private let signer: any JwsSigner
    private let jwk: JsonValue
    private let now: () -> Int64

    public init(signer: any JwsSigner, publicKey: EcPublicKey, now: @escaping () -> Int64) {
        self.signer = signer
        self.jwk = JwkEc.toJson(publicKey)
        self.now = now
    }

    public func proofJwt(credentialIssuer: String, cNonce: String?, clientId: String? = nil) async throws -> String {
        let header = JsonValue.obj([
            ("typ", .str("openid4vci-proof+jwt")),
            ("alg", .str(jwsAlgName(signer.algorithm))),
            ("jwk", jwk),
        ])
        var claims: [(String, JsonValue)] = []
        if let clientId { claims.append(("iss", .str(clientId))) }
        claims.append(("aud", .str(credentialIssuer)))
        claims.append(("iat", .numInt(now())))
        if let cNonce { claims.append(("nonce", .str(cNonce))) }
        let jws = try await Jws.sign(header: header, payload: [UInt8](JsonValue.obj(claims).serialize().utf8), signer: signer)
        return jws.compact()
    }
}

File: kotlin/openid4vci/src/main/kotlin/com/hopae/eudi/wallet/vci/Crypto.kt
package com.hopae.eudi.wallet.vci

import com.hopae.eudi.wallet.sdjwt.Base64Url
import com.hopae.eudi.wallet.sdjwt.JsonValue
import com.hopae.eudi.wallet.sdjwt.JwkEc
import com.hopae.eudi.wallet.sdjwt.Jws
import com.hopae.eudi.wallet.sdjwt.JwsSigner
import com.hopae.eudi.wallet.spi.Rng
import java.security.MessageDigest

internal fun sha256(bytes: ByteArray): ByteArray = MessageDigest.getInstance("SHA-256").digest(bytes)

/** PKCE (RFC 7636) with the S256 method — the only method HAIP permits. */
class Pkce private constructor(val codeVerifier: String, val codeChallenge: String) {
    val method: String get() = "S256"

    companion object {
        fun create(rng: Rng): Pkce {
            val verifier = Base64Url.encode(rng.nextBytes(32))
            val challenge = Base64Url.encode(sha256(verifier.encodeToByteArray()))
            return Pkce(verifier, challenge)
        }
    }
}

/**
 * DPoP proof builder (RFC 9449). One instance per proof-key; produces a fresh proof per
 * request (unique jti, correct htm/htu, optional ath binding to the access token, and a
 * server-provided nonce on retry).
 */
class DpopProver(
    private val signer: JwsSigner,
    publicKey: com.hopae.eudi.wallet.cbor.cose.EcPublicKey,
    private val rng: Rng,
    private val now: () -> Long,
) {
    private val jwk: JsonValue.Obj = JwkEc.toJson(publicKey)

    suspend fun proof(
        method: String,
        url: String,
        accessToken: String? = null,
        nonce: String? = null,
    ): String {
        val header = JsonValue.Obj(
            listOf(
                "typ" to JsonValue.Str("dpop+jwt"),
                "alg" to JsonValue.Str(jwsAlg()),
                "jwk" to jwk,
            )
        )
        val claims = mutableListOf<Pair<String, JsonValue>>(
            "jti" to JsonValue.Str(Base64Url.encode(rng.nextBytes(16))),
            "htm" to JsonValue.Str(method),
            "htu" to JsonValue.Str(htu(url)),
            "iat" to JsonValue.NumInt(now()),
        )
        accessToken?.let { claims.add("ath" to JsonValue.Str(Base64Url.encode(sha256(it.encodeToByteArray())))) }
        nonce?.let { claims.add("nonce" to JsonValue.Str(it)) }
        return Jws.sign(header, JsonValue.Obj(claims).serialize().encodeToByteArray(), signer).compact()
    }

    private fun jwsAlg(): String = when (signer.algorithm) {
        com.hopae.eudi.wallet.spi.SigningAlgorithm.ES256 -> "ES256"
        com.hopae.eudi.wallet.spi.SigningAlgorithm.ES384 -> "ES384"
        com.hopae.eudi.wallet.spi.SigningAlgorithm.ES512 -> "ES512"
    }

    /** htu is the request URI without query and fragment (RFC 9449 §4.2). */
    private fun htu(url: String): String = url.substringBefore('?').substringBefore('#')
}

/**
 * OpenID4VCI key-proof-of-possession builder (§8.2.1): typ `openid4vci-proof+jwt`, the
 * holder's public JWK in the header, `aud`=issuer, and the issuer `c_nonce`.
 */
class KeyProofSigner(
    private val signer: JwsSigner,
    publicKey: com.hopae.eudi.wallet.cbor.cose.EcPublicKey,
    private val now: () -> Long,
) {
    private val jwk: JsonValue.Obj = JwkEc.toJson(publicKey)

    suspend fun proofJwt(credentialIssuer: String, cNonce: String?, clientId: String? = null): String {
        val header = JsonValue.Obj(
            listOf(
                "typ" to JsonValue.Str("openid4vci-proof+jwt"),
                "alg" to JsonValue.Str(
                    when (signer.algorithm) {
                        com.hopae.eudi.wallet.spi.SigningAlgorithm.ES256 -> "ES256"
                        com.hopae.eudi.wallet.spi.SigningAlgorithm.ES384 -> "ES384"
                        com.hopae.eudi.wallet.spi.SigningAlgorithm.ES512 -> "ES512"
                    }
                ),
                "jwk" to jwk,
            )
        )
        val claims = mutableListOf<Pair<String, JsonValue>>(
            "aud" to JsonValue.Str(credentialIssuer),
            "iat" to JsonValue.NumInt(now()),
        )
        clientId?.let { claims.add(0, "iss" to JsonValue.Str(it)) }
        cNonce?.let { claims.add("nonce" to JsonValue.Str(it)) }
        return Jws.sign(header, JsonValue.Obj(claims).serialize().encodeToByteArray(), signer).compact()
    }
}
